### ISSUE TITLE
Anyone can execute migration once migration destination is set

### ISSUE THREAD

**Issue by zanderbytexyz** [shieldify-security/Adapters-SPP#13]
## Severity
High
## Proof of Concept (PoC)
```solidity
    /// @notice This function mints the Portal NFT and transfers user stakes to a new Adapter
    /// @dev Timelock protected function that can only be called once to move capital to a new Adapter
    function executeMigration() external isMigrating {
        /// @dev Ensure that the timelock has passed
        if (block.timestamp < migrationTime) {
            revert ErrorsLib.isTimeLocked();
        }

        /// @dev Ensure that the migration (minting of NFT) can only be performed once
        if (successMigrated == true) {
            revert ErrorsLib.hasMigrated();
        }

        /// @dev Mint an NFT to the new Adapter that holds the current Adapter stake information
        /// @dev IMPORTANT: The migration contract must be able to receive ERC721 tokens
        successMigrated = true;
        totalPrincipalStaked = 0;
        PORTAL.mintNFTposition(migrationDestination);

        /// @dev Emit migration event
        emit EventsLib.migrationExecuted(migrationDestination);
    }
```
## Recommendation

**Comment by PossumLabsCrypto** [comment]
> migrationTime equals 0, it is only set when this condition is true in acceptMigrationDestination function:

```
/// @dev Ensure that the timelock has passed
        if (block.timestamp < migrationTime && migrationTime > 0) {
            revert ErrorsLib.isTimeLocked();
        }
```

The second part of the check, migrationTime > 0 prevents the execution without votes.

**Comment by zanderbytexyz** [comment]
Sorry but I do not agree with this statement.
The claim that "the second part of the check, migrationTime > 0, prevents the execution without votes" is simply wrong. 
Here’s the clarification:
We revert only if `migrationTime > 0`, meaning if `migrationTime = 0` the check passes without issue. I’ve provided PoC above (after making the tests work) and you can test it yourself. 
Again the flow is as follows:
1. Owner propose migrationDestination
`migrationTime = 0` (by default). It is only initialized when `votesForMigration > totalParticipalStaked / 2`
The check only revert if migrationDestination is set -> `block.timestamp < migrationTime` this evaluates to false, we have also `&&`, `migrationTime > 0`, since the first one is evaluated to false we don’t even go to the second check and also it will evaluate to true only when migrationTime is set.
2. `executeMigration()` is called (by anyone) and executed.

My suggestion is to split the checks into two separate ones:
```solidity
 if(migrationTime == 0) revert ErrorsLib.CustomError();
```
And

```solidity
if (block.timestamp < migrationTime) {
            revert ErrorsLib.isTimeLocked();
        }
```

**Comment by PossumLabsCrypto** [comment]
Code blindness - you´re right. 👍 

Acknowledging this as medium like the original report #9 and rejudging the original report to low because the reviewer didn´t catch the bug in the fix which causes additional expenses.

**Comment by zanderbytexyz** [comment]
Thank you for your response! I have some opinions on the topic which I want to express:
I think this is a valid high finding and these are my arguments:
* The issue directly broke one of the contract core functionalities 
* Anyone cal call that function and pass migration even if they don’t stake a single wei
* Voting becomes meaningless since the moment a new `migrationDestinations` is proposed, it can be accepted even if users disagree
* There is an indirect loss for the protocol, as users losing their voting rights aren’t incentivized to stake and use the protocol
* Even existing users who have already staked may lose trust in the protocol if this vulnerability is exploited
* Regarding the question - “Do we need to have one staked balance equal to one vote?”, the team answers as follows - “Yes to avoid Sybil attacks. **The majority of capital should be in favor of migration, otherwise it shouldn’t pass.** More capital = more incentive to actually check the code of the destination address.” and here that is not the case.

**Comment by PossumLabsCrypto** [comment]
Nope, this is not a high. There is no loss of funds and no loss for the protocol since the capital is staked in the underlying Portal which constitutes "the protocol".

Migration is not a core functionality. In fact, it´s entirely optional since users can migrate manually to any other contract by unstaking and restaking. The migration function only exists for the convenience of users.

**Comment by zanderbytexyz** [comment]
I understand you. Thanks!

**Comment by PossumLabsCrypto** [comment]
[fix applied](https://github.com/PossumLabsCrypto/Adapters/commit/e300db0831783de9971101d4ff9d87eaceacfde5) 

this is your suggested fix in shorter form.
```
        /// @dev Ensure that the timelock is set and has passed
        if (block.timestamp < migrationTime || migrationTime == 0) {
            revert ErrorsLib.isTimeLocked();
        }
```

Your suggestion:

```
 if(migrationTime == 0) revert ErrorsLib.CustomError();
 if (block.timestamp < migrationTime) {
            revert ErrorsLib.isTimeLocked();
        }
```

**Comment by zanderbytexyz** [comment]
I've applied the changes above and tested it. Looks good to me. 
```solidity
function testExecuteMigrationWithoutVotes() external {
        help_stake_ETH();
        help_stake_ETH_Bob();
        help_stake_ETH_Karen();

        vm.prank(owner);
        adapter_ETH.proposeMigrationDestination(newAdapterAddr);

        vm.prank(address(1));
        vm.expectRevert(ErrorsLib.isTimeLocked.selector);
        adapter_ETH.executeMigration();
    }
```
